# Schnittpunkt und Schnittwinkel zweier Geraden

In [3]:
# Bibliotheken

import numpy as np
import numpy.linalg as la
import matplotlib.pyplot as pl
import ipywidgets as iw

%matplotlib widget

ModuleNotFoundError: No module named 'matplotlib'

## Geraden definieren

### 4 Punkte

<a href="https://de.wikipedia.org/wiki/Liste_der_Flugh%C3%A4fen_in_der_Schweiz">Standorte der 11 Schweizer Regionalflugplätze</a>
<br/>
Koordinaten in Schweizer Landeskoordinaten CH1903/LV03 (veraltet): 
<a href="https://de.wikipedia.org/wiki/Schweizer_Landeskoordinaten">Wikipedia</a> /
<a href="https://www.swisstopo.admin.ch/de/schweizer-koordinatensystem">swisstopo</a>
<br/>
<a href="https://map.geo.admin.ch">Link zur Online-Karte von swisstopo</a>

#### 1. Flugplatz Lausanne - La Blécherette

<a href="https://geohack.toolforge.org/geohack.php?pagename=Liste_der_Flugh%C3%A4fen_in_der_Schweiz&language=de&params=46.545277777778_N_6.6166666666667_E_region:CH_type:landmark&title=%5B%5BFlugplatz+Lausanne-La+Bl%C3%A9cherette%7CLausanne%5D%5D">Wikipedia</a>
/
<a href="https://s.geo.admin.ch/a1ce2e0bcd">swisstopo</a>

Als Vektor in mathematischer Schreibweise

$ \mathrm{p} = \begin{pmatrix} 536.955 \\ 155.219 \end{pmatrix} $

In Python / NumPy

In [4]:
p = np.array ([
    [536.955],
    [155.219]
])

#### 2. Flugplatz St. Gallen - Altenrhein

<a href="https://geohack.toolforge.org/geohack.php?pagename=Liste_der_Flugh%C3%A4fen_in_der_Schweiz&language=de&params=47.485027777778_N_9.5607777777778_E_region:CH_type:landmark&title=%5B%5BFlugplatz+St.+Gallen-Altenrhein%7CSt.+Gallen-Altenrhein%5D%5D">Wikipedia</a>
/
<a href="https://s.geo.admin.ch/a1ce2ec299">swisstopo</a>

In [5]:
r = np.array ([
    [759.925],
    [261.525]
])

#### 3. Flugplatz Birrfeld

<a href="https://geohack.toolforge.org/geohack.php?pagename=Liste_der_Flugh%C3%A4fen_in_der_Schweiz&language=de&params=47.443333333333_N_8.2338888888889_E_region:CH_type:landmark&title=%5B%5BFlugplatz+Birrfeld%7CBirrfeld%5D%5D">Wikipedia</a>
/
<a href="https://s.geo.admin.ch/a1cdd71ce9">swisstopo</a>

In [6]:
q = np.array ([
    [659.984],
    [255.030]
])

#### 4. Flughafen Lugano

<a href="https://geohack.toolforge.org/geohack.php?pagename=Liste_der_Flugh%C3%A4fen_in_der_Schweiz&language=de&params=46.004166666667_N_8.9105555555556_E_region:CH_type:landmark&title=%5B%5BFlughafen+Lugano%7CLugano%5D%5D">Wikipedia</a>
/
<a href="https://s.geo.admin.ch/a1cdd9c785">swisstopo</a>

In [7]:
s = np.array ([
    [714.020],
    [95.804]
])

### 2 Geraden

#### 1. Flug von Blecherette nach Altenrhein

$ \mathrm{g} \, (y) \; = \; (1 - y) \, \mathrm{p} + y \, \mathrm{r} \;; \qquad y \in [0, 1] $

##### Mathematisch als Gerade in Punkt-Richtungs-Form

$ g
\; = \; \mathrm{p} - y \, \mathrm{p} + y \, \mathrm{r}
\; = \; \mathrm{p} + y \, (\mathrm{r} - \mathrm{p})
\; = \; \mathrm{p} + y \, \mathrm{u} \;; \qquad \mathrm{u} \; = \; \mathrm{r} - \mathrm{p}
$

##### Richtungsvektor in Python

In [8]:
u = r - p
print (u)

[[222.97 ]
 [106.306]]


##### Geradengleichung

$
\mathrm{g} \, (y)
\; = \; \begin{pmatrix} 536.955 \\ 155.219 \end{pmatrix}
+ y \, \begin{pmatrix} 222.970 \\ 106.306 \end{pmatrix}
$

#### 2. Flug von Birrfeld nach Lugano

In [9]:
v = s - q
print (v)

[[  54.036]
 [-159.226]]


##### Geradengleichung

$
\mathrm{h} \, (z)
\; = \; \begin{pmatrix} 659.984 \\ 255.030 \end{pmatrix}
+ z \, \begin{pmatrix} 54.036 \\ -159.226 \end{pmatrix}
$

#### Beide Flugrouten zeichnen

In [10]:
def PlotVec (p, u, c):
    pl.arrow (p [0] [0], p [1] [0], u [0] [0], u [1] [0],
        color = c,
        head_width = 5,
        length_includes_head = True,
        fill = False)

In [12]:
PlotVec (p, u, 'blue')
PlotVec (q, v, 'green')
pl.axis ('equal')
pl.grid ()
pl.show ()

NameError: name 'pl' is not defined

#### Schnittpunkt berechnen

##### Experimentell (interaktiv)

In [ ]:
def PlotPnt (p, c):
    return pl.plot (p [0], p [1], 'o', color = c)

In [ ]:
def Interaktiv1 ():

    fig, axs = pl.subplots()
    cnv = fig.canvas
    cnv.toolbar_visible = False
    cnv.header_visible = False
    cnv.footer_visible = False
    
    PlotVec (p, u, 'blue' ); P, = PlotPnt (p, 'blue' )
    PlotVec (q, v, 'green'); Q, = PlotPnt (q, 'green')
    pl.axis ('equal')
    pl.grid ()
    pl.show()

    def Callback1 (y, z):
        g = p + y * u; P.set_data (g [0], g [1])
        h = q + z * v; Q.set_data (h [0], h [1])
    
    Slider11 = iw.FloatSlider (min = -0.1, max = 1.1, step = 0.001, value = 0.5, description = 'y [%]')
    Slider12 = iw.FloatSlider (min = -0.1, max = 1.1, step = 0.001, value = 0.5, description = 'z [%]')
    UserInterface1 = iw.HBox ([Slider11, Slider12])
    Output1 = iw.interactive_output (Callback1, {'y': Slider11, 'z': Slider12})
    
    display (UserInterface1, Output1)

In [ ]:
Interaktiv1()

##### Berechnung (Mathematik)

###### vektoriell

$ \mathrm{g} \, (y) = \mathrm{h} \, (z) $

$ p + y \, \mathrm{u} = q + z \, \mathrm{v} $

$
y \, \mathrm{u} - z \, \mathrm{v}
    \; = \; \mathrm{q} - \mathrm{p}
$

$
y \, \begin{pmatrix} u_1 \\ u_2 \end{pmatrix} - z \, \begin{pmatrix} v_1 \\ v_2 \end{pmatrix}
    \; = \; \begin{pmatrix} y \, u_1 \\ y \, u_2 \end{pmatrix} - \begin{pmatrix} z \, v_1 \\ z \, v_2 \end{pmatrix}
    \; = \; \begin{pmatrix} y \, u_1 - z \, v_1 \\ z \, u_2 - z \, v_2 \end{pmatrix}
    \quad = \quad \begin{pmatrix} q_1 \\ q_2 \end{pmatrix} - \begin{pmatrix} p_1 \\ p_2 \end{pmatrix}
    \; = \; \begin{pmatrix} q_1 - p_1 \\ q_2 - p_2 \end{pmatrix}
$

###### als 2 x 2 - Gleichungssystem

$\begin{align}
u_1 \, y - v_1 \, z &\;=\; q_1 - p_1 \\
u_2 \, y - v_2 \, z &\;=\; q_2 - p_2
\end{align}$

###### in Matrixform

$
\begin{pmatrix} u_1 \, y - v_1 \, z \\ u_2 \, y - v_2 \, z \end{pmatrix}
    \; = \; \begin{pmatrix} u_1 & - \, v_1 \\ u_2 & - v_2 \end{pmatrix}
        \begin{pmatrix} y \\ z \end{pmatrix}
    \; = \; \begin{pmatrix} q_1 - p_1 \\ q_2 - p_2 \end{pmatrix}
$

$
    \mathrm{A} \, \mathrm{x} \; = \; \mathrm{b} \;;
        \qquad \mathrm{A} \; = \; \begin{pmatrix} u_1 & - v_1 \\ u_2 & - v_2\end{pmatrix} \;;
        \qquad \mathrm{x} \; = \; \begin{pmatrix} y \\ z \end{pmatrix} \;;
        \qquad \mathrm{b} \; = \; \begin{pmatrix} q_1 - p_1 \\ u_2 - p_2 \end{pmatrix}
$

##### Berechnung (Python)

In [ ]:
A = np.concatenate ((u, -v), axis = 1)
print (A)

In [ ]:
b = q - p
print (b)

In [ ]:
x = la.solve (A, b)
print (x)

In [ ]:
[y, z] = x
print (y)
print (z)

In [ ]:
g = p + y * u
print (g)

##### Probe

In [ ]:
h = q + z * v
print (h)

In [ ]:
g - h

##### Zeichnung

In [ ]:
PlotVec (p, u, 'blue')
PlotVec (q, v, 'green')
PlotPnt (g, 'red')
pl.axis ('equal')
pl.grid ()
pl.show ()

##### <a href="https://s.geo.admin.ch/a1cdca3cdb">Der Schnittpunkt auf der Karte</a>

## Geraden parametrisiert über die zurückgelegte Strecke

### Länge der Strecken

In [ ]:
Lu = la.norm (u)
print (Lu)

In [ ]:
Lv = la.norm (v)
print (Lv)

### Richtungsvektoren skalieren

In [ ]:
u0 = u / Lu
print (u0)

In [ ]:
v0 = v / Lv
print (v0)

#### Jeder Richtungsvektor ist jetzt 1 km lang

In [ ]:
la.norm (u0)

In [ ]:
la.norm (v0)

### Interaktive Zeichnung

In [ ]:
def Interaktiv2 ():

    fig, axs = pl.subplots()
    cnv = fig.canvas
    cnv.toolbar_visible = False
    cnv.header_visible = False
    cnv.footer_visible = False

    PlotVec (p, u, 'blue' ); P, = PlotPnt (p, 'blue' )
    PlotVec (q, v, 'green'); Q, = PlotPnt (q, 'green')
    pl.axis ('equal')
    pl.grid ()
    pl.show()
    
    def Callback2 (y, z):
        pu = p + y * u0; P.set_data (pu [0], pu [1])
        qv = q + z * v0; Q.set_data (qv [0], qv [1])
        
    Slider21 = iw.FloatSlider (min = 0.0, max = Lu, step = 0.1, value = Lu / 2.0, description = 'y [km]')
    Slider22 = iw.FloatSlider (min = 0.0, max = Lv, step = 0.1, value = Lv / 2.0, description = 'z [km]')
    UserInterface2 = iw.HBox ([Slider21, Slider22])
    Output2 = iw.interactive_output (Callback2, {'y': Slider21, 'z': Slider22})
    
    display (UserInterface2, Output2)

Interaktiv2()

## Geraden parametrisiert über die Zeit

### Gerade g: Lausanne Blecherette - St. Gallen Altenrhein

<a href="https://de.wikipedia.org/wiki/Piper_PA-28">Piper PA-28</a>

Reisegeschwindigkeit 100&hellip;140 kn = 180&hellip;260 km/h

### Gerade h: Birrfeld - Lugano

<a href="https://de.wikipedia.org/wiki/Cessna_172">Cessna 172</a>

Reisegeschwindigkeit 120 kn = 222 km/h

### Interaktive Zeichnung

In [ ]:
def Interaktiv3 ():

    fig, axs = pl.subplots()
    cnv = fig.canvas
    cnv.toolbar_visible = False
    cnv.header_visible = False
    cnv.footer_visible = False

    PlotVec (p, u, 'blue' ); P, = PlotPnt (p, 'blue' )
    PlotVec (q, v, 'green'); Q, = PlotPnt (q, 'green')
    pl.axis ('equal')
    pl.grid ()
    pl.show()

    def Callback3 (time, speed):
        u1 = u0 * speed; pu = p + time * u1; P.set_data (pu [0], pu [1])
        v1 = v0 * 222.0; qv = q + time * v1; Q.set_data (qv [0], qv [1])
    
    Slider31 = iw.FloatSlider (min = 0.0, max = 1.2, step = 0.001, value = 0.0, description = 't [h]')
    Slider32 = iw.FloatSlider (min = 180.0, max = 2600.0, step = 1, value = (180.0 + 260.0)  / 2.0, description = 'v [km/h]')
    UserInterface3 = iw.HBox ([Slider31, Slider32])
    Output3 = iw.interactive_output (Callback3, {'time': Slider31, 'speed': Slider32})
    
    display (UserInterface3, Output3)

Interaktiv3()

Kollision bei einer Geschwindigkeit von ca. 885 km/h auf der Strecke Lausanne - St. Gallen <br/>
z. B. mit einem <a href="https://de.wikipedia.org/wiki/Airbus_A330">Airbus A330-200</a>

## Schnittwinkel

### Skalarprodukt

In [ ]:
s = np.sum (u * v)
print (s)

### Quotient aus Skalarprodukt und Produkt der Längen ergibt den Cosinus

In [ ]:
c = s / (Lu * Lv)
print (c)

### Winkel

#### im Bogenmass

In [ ]:
b = np.arccos (c)
print (b)

#### und im Gradmass

In [ ]:
w1 = b * 180.0 / np.pi
print (w1)

Komplementärwinkel im Gradmass

In [ ]:
w2 = 180 - w1
print (w2)